modules/test-operators/apis/keystone.go

```diff
@@ -0,0 +1,303 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apis
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/domains"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/users"
+	"golang.org/x/exp/maps"
+)
+
+type KeystoneAPIFixture struct {
+	APIFixture
+	// Users is the map of user objects known by the fixture keyed by the user
+	// name
+	Users map[string]users.User
+	// Domains is the map of domain objects known by the fixture keyed by the
+	// domain name
+	Domains map[string]domains.Domain
+}
+
+// Sets up a keystone-api simulator with an embedded http server
+func NewKeystoneAPIFixtureWithServer(log logr.Logger) *KeystoneAPIFixture {
+	server := &FakeAPIServer{}
+	server.Setup(log)
+	fixture := AddKeystoneAPIFixture(log, server)
+	fixture.ownsServer = true
+	return fixture
+}
+
+// Adds a keystone-api simulator to an already created http server.
+// This is useful if you need to have more than one openstack API simulated
+// but you don't want to start a separate http server for each.
+func AddKeystoneAPIFixture(log logr.Logger, server *FakeAPIServer) *KeystoneAPIFixture {
+	fixture := &KeystoneAPIFixture{
+		APIFixture: APIFixture{
+			server:     server,
+			log:        log,
+			urlBase:    "/identity",
+			ownsServer: false,
+		},
+		Users:   map[string]users.User{},
+		Domains: map[string]domains.Domain{},
+	}
+	return fixture
+}
+
+// Setup adds the API request handlers to the fixture. If no handlers is passed
+// then a basic set of well behaving handlers are added that will simulate the
+// happy path.
+// If you need to customize the behavior of the fixture, e.g. to inject faults,
+// then you can pass a list of handlers to register instead.
+func (f *KeystoneAPIFixture) Setup(handlers ...Handler) {
+	if len(handlers) == 0 {
+		f.registerNormalHandlers()
+	}
+	for _, handler := range handlers {
+		f.registerHandler(handler)
+	}
+}
+
+func (f *KeystoneAPIFixture) registerNormalHandlers() {
+	f.registerHandler(Handler{Patter: "/", Func: f.HandleVersion})
+	f.registerHandler(Handler{Patter: "/v3/auth/tokens", Func: f.HandleToken})
+	f.registerHandler(Handler{Patter: "/v3/users", Func: f.HandleUsers})
+	f.registerHandler(Handler{Patter: "/v3/domains", Func: f.HandleDomains})
+}
+
+func (f *KeystoneAPIFixture) registerHandler(handler Handler) {
+	f.server.mux.HandleFunc(f.urlBase+handler.Patter, handler.Func)
+}
+
+func (f *KeystoneAPIFixture) HandleVersion(w http.ResponseWriter, r *http.Request) {
```

Comment:
IMO adding some doc strings here and in the functions below would be a nice touch

Reply:
Done